libs/akita/src/lib/entityStore.ts

```diff
@@ -204,28 +218,36 @@ export class EntityStore<S extends EntityState = any, EntityType = getEntityType
    *
    * @example
    *
-   * store.upsert(1, { active: true })
-   * store.upsert([2, 3], { active: true })
-   * store.upsert([2, 3], entity => ({ isOpen: !entity.isOpen}))
+   * store.upsert(1, { active: true }, (id, newState) => ({ id, ...newState, enabled: true }))
+   * store.upsert([2, 3], { active: true }, (id, newState) => ({ id, ...newState, enabled: true }))
+   * store.upsert([2, 3], entity => ({ isOpen: !(entity?.isOpen ?? true) }), (id, newState) => ({ id, ...newState, enabled: true }))
    *
    */
   @transaction()
-  upsert(ids: OrArray<IDType>, newState: Partial<EntityType> | EntityType | UpdateStateCallback<EntityType> | EntityType[], options: { baseClass?: Constructor } = {}) {
+  upsert<NewEntityType extends Partial<EntityType>>(
+    ids: OrArray<IDType>,
+    newState: UpsertStateCallback<EntityType, NewEntityType> | NewEntityType,
+    onCreate: CreateStateCallback<EntityType, NewEntityType, IDType>,
+    options: { baseClass?: Constructor } = {}
+  ) {
     const toArray = coerceArray(ids);
-    const predicate = isUpdate => id => hasEntity(this.entities, id) === isUpdate;
-    const isClassBased = isFunction(options.baseClass);
+    const predicate = (isUpdate) => (id) => hasEntity(this.entities, id) === isUpdate;
+    const baseClass = options.baseClass;
+    const isClassBased = isFunction(baseClass);
     const updateIds = toArray.filter(predicate(true));
-    const newEntities = toArray.filter(predicate(false)).map(id => {
-      let entity = isFunction(newState) ? newState({} as EntityType) : newState;
+    const newEntities = toArray.filter(predicate(false)).map((id) => {
+      const oldStateObj = this.entities[id as any];
+      const newStateObj = typeof newState === 'function' ? newState(oldStateObj) : newState;
+      const entity = onCreate(id, newStateObj ? newStateObj : undefined);
```

Comment:
In your examples, it seems like you always have the `newState` param. But in reality, it can be `undefined`.

Reply:
I missed that when I removed the backward compatibility. It can never be `undefined`, because the previous update callback always returns an object.